ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -62,16 +72,51 @@ def get_waypoint_velocity(self, waypoint):
     def set_waypoint_velocity(self, waypoints, waypoint, velocity):
         waypoints[waypoint].twist.twist.linear.x = velocity
 
-    def distance(self, waypoints, wp1, wp2):
-        dist = 0
-        dl = lambda a, b: math.sqrt((a.x-b.x)**2 + (a.y-b.y)**2  + (a.z-b.z)**2)
-        for i in range(wp1, wp2+1):
-            dist += dl(waypoints[wp1].pose.pose.position, waypoints[i].pose.pose.position)
-            wp1 = i
-        return dist
+    def distance_to_previous(self, position):
+        return euclidean(get_position(self.previous_pos), get_position(position))
+
+    def waypoint_is_ahead(self, waypoint):
+        w_x, w_y, _ = get_position(waypoint.pose.pose)
+        p_x, p_y, _ = get_position(self.previous_pos)
+        heading = math.atan2(w_y-p_y, w_x-p_x)
+        yaw = euler_from_quaternion(get_orientation(self.previous_pos))[2]
+        angle = yaw-heading
+        return True if math.fabs(angle) < math.pi/4. else False
+
+    def update_next_waypoint(self):
+        it = 0
+        nb_waypoints = len(self.static_waypoints.waypoints)
+        while not self.waypoint_is_ahead(self.static_waypoints.waypoints[self.next_waypoint % nb_waypoints]) and \
+                it < UPDATE_MAX_ITER:
+            self.next_waypoint += 1  # We look at the next one
+
+        # Searching the next waypoint in the full path takes much longer, we want to avoid it as much as possible
+        if it == UPDATE_MAX_ITER:
+            self.search_next_waypoint()
+
+    def search_next_waypoint(self):
+        self.next_waypoint = 0
+        # TODO: Rewrite this in a clearer way
+        # We basically search among all static waypoints the closest waypoint ahead
+        for i in range(len(self.static_waypoints)):
+            if self.waypoint_is_ahead(self.static_waypoints.waypoints[i]) and \
+                    self.distance_to_previous(self.static_waypoints.waypoints[i].pose) < \
+                    self.distance_to_previous(self.static_waypoints.waypoints[self.next_waypoint].pose):
+                self.next_waypoint = i
```

Comment:
probably you can brake the loop when you find what the next waypoint.

Reply:
But how do I know that this is the closest waypoint ahead if I stop early ?